nilearn/surface/surface.py

```diff
@@ -1200,39 +1045,99 @@ def check_mesh_and_data(mesh, data):
 
     Returns
     -------
-    surface : Surface
-        Checked surface object.
+    mesh : SurfaceMesh
+        With the fields "coordinates" and "faces", each containing a
+        :obj:`numpy.ndarray`
 
     """
-    surface = load_surface(surface)
-    mesh, data = check_mesh_and_data(surface.mesh, surface.data)
-    return Surface(mesh, data)
+    # if input is a filename, try to load it
+    surf_mesh = stringify_path(surf_mesh)
+    if isinstance(surf_mesh, str):
+        # resolve globbing
+        file_list = resolve_globbing(surf_mesh)
+        if len(file_list) > 1:
+            # empty list is handled inside resolve_globbing function
+            raise ValueError(
+                f"More than one file matching path: {surf_mesh}\n"
+                "load_surf_mesh can only load one file at a time."
+            )
+        surf_mesh = str(file_list[0])
+
+        if any(surf_mesh.endswith(x) for x in FREESURFER_MESH_EXTENSIONS):
+            coords, faces, header = fs.io.read_geometry(
+                surf_mesh, read_metadata=True
+            )
+            # See https://github.com/nilearn/nilearn/pull/3235
+            if "cras" in header:
+                coords += header["cras"]
+            mesh = InMemoryMesh(coordinates=coords, faces=faces)
+        elif surf_mesh.endswith("gii"):
+            coords, faces = _gifti_img_to_mesh(load(surf_mesh))
+            mesh = InMemoryMesh(coordinates=coords, faces=faces)
+        elif surf_mesh.endswith("gii.gz"):
+            gifti_img = _load_surf_files_gifti_gzip(surf_mesh)
+            coords, faces = _gifti_img_to_mesh(gifti_img)
+            mesh = InMemoryMesh(coordinates=coords, faces=faces)
+        else:
+            raise ValueError(
+                "The input type is not recognized. "
+                f"{surf_mesh!r} was given "
+                "while valid inputs are one of the following "
+                "file formats: .gii, .gii.gz, "
+                "Freesurfer specific files such as "
+                f"{_stringify(FREESURFER_MESH_EXTENSIONS)}, "
+                "two Numpy arrays organized in a list, tuple "
+                "or a namedtuple with the "
+                'fields "coordinates" and "faces".'
+            )
+    elif isinstance(surf_mesh, (list, tuple)):
+        try:
+            coords, faces = surf_mesh
+            mesh = InMemoryMesh(coordinates=coords, faces=faces)
+        except Exception:
+            raise ValueError(
+                "If a list or tuple is given as input, "
+                "it must have two elements, the first is "
+                "a Numpy array containing the x-y-z coordinates "
+                "of the mesh vertices, the second is a Numpy "
+                "array containing  the indices (into coords) of "
+                "the mesh faces. The input was a list with "
+                f"{len(surf_mesh)} elements."
+            )
+    elif hasattr(surf_mesh, "faces") and hasattr(surf_mesh, "coordinates"):
```

Comment:
why not check directly if the type is InMemoryMesh or SurfaceMesh?

Reply:
As it turns out, you cannot instantiate a SurfaceMesh object with coordinates and faces. 
It only works with InMemoryMesh.

The issue is that `load_surf_mesh` allows users to load a single hemisphere. It seems to me that the whole point of the new surface API was to discourage users from doing that and instead force them to specify which hemisphere they want to load/define.

However, we use this function internally within InMemoryMesh, and it does make sense to use it there. So in the long term, we should probably remove this function from public API and only use it privately.

Reply:
Same applies for `load_surf_data`

Reply:
PolyMesh and PolyData or simply dicts seem to be the objects we want users to be using instead.

Reply:
ok thanks for the clarification